cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/executor/strategy/AbstractReplicaMovementStrategy.java

```diff
@@ -0,0 +1,84 @@
+
+import com.linkedin.kafka.cruisecontrol.executor.ExecutionProposal;
+import com.linkedin.kafka.cruisecontrol.executor.ExecutionTask;
+import java.util.Collection;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.SortedSet;
+import java.util.TreeSet;
+import org.apache.kafka.common.Cluster;
+
+
+/**
+ * An abstract class for replica movement strategy. This class will be extended to create custom strategy to determine the
+ * execution order the replica movement tasks.
+ */
+public abstract class AbstractReplicaMovementStrategy implements ReplicaMovementStrategy {
+
+  /**
+   * Generate a comparator for replica movement task which incorporate the strategy to apply. The "smaller" task will have
+   * higher execution priority.
+   *
+   * @param cluster The current cluster state.
+   * @return The comparator of task.
+   */
+  public abstract Comparator<ExecutionTask> taskComparator(Cluster cluster);
+
+  @Override
+  public ReplicaMovementStrategy chain(ReplicaMovementStrategy strategy) {
+    if (!(strategy instanceof AbstractReplicaMovementStrategy)) {
+      throw new IllegalArgumentException("Replica movement strategy " + this.getClass().getSimpleName() + " is unable to chain with "
+                                         + strategy.getClass().getSimpleName() + ".");
+    }
+    AbstractReplicaMovementStrategy current = this;
+    return new AbstractReplicaMovementStrategy() {
+      @Override
+      public Comparator<ExecutionTask> taskComparator(Cluster cluster) {
+        Comparator<ExecutionTask> comparator1 = current.taskComparator(cluster);
+        Comparator<ExecutionTask> comparator2 = ((AbstractReplicaMovementStrategy) strategy).taskComparator(cluster);
+
+        return (task1, task2) -> {
+          int compareResult1 = comparator1.compare(task1, task2);
+          int compareResult2 = comparator2.compare(task1, task2);
+          if (compareResult1 == 0) {
+            return compareResult2;
+          }
+          return compareResult1;
+        };
+      }
+    };
+  }
+
+  @Override
+  public Map<Integer, SortedSet<ExecutionTask>> applyStrategy(Collection<ExecutionTask> replicaMovementTasks, Cluster cluster) {
+    Map<Integer, SortedSet<ExecutionTask>> tasksByBrokerId = new HashMap<>();
+
+    for (ExecutionTask task : replicaMovementTasks) {
+      ExecutionProposal proposal = task.proposal();
+
+      // Add the task to source broker's execution plan
+      int sourceBroker = proposal.oldLeader();
+      SortedSet<ExecutionTask> sourceBrokerTaskSet = tasksByBrokerId.computeIfAbsent(sourceBroker,
+                                                                                     k -> new TreeSet<>(taskComparator(cluster)));
+      if (!sourceBrokerTaskSet.add(task)) {
+        throw new IllegalStateException("Replica movement strategy " + this.getClass().getSimpleName() + " is unable to determine order of all tasks.");
```

Comment:
Will `this.getClass().getSimpleName()` return `AbstractReplicaMovementStrategy` rather than the class name that implements it?

Reply:
I think it will return the the implemented class type since `Object.getClass()` returns the runtime class of the object itself.

Reply:
The sanity check would be redundant once we ensure the use of `BaseReplicaMovementStrategy` (please see previous discussion).